# Introduction

Do higher film budgets lead to more box office revenue? Let's find out if there's a relationship using the movie budgets and financial performance data that I've scraped from [the-numbers.com](https://www.the-numbers.com/movie/budgets) on **May 1st, 2018**. 

<img src=https://i.imgur.com/kq7hrEh.png>

# Import Statements

In [47]:
import pandas as pd
import matplotlib.pyplot as plt
import math


# Notebook Presentation

In [2]:
pd.options.display.float_format = '{:,.2f}'.format

from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()

# Read the Data

In [3]:
data = pd.read_csv('cost_revenue_dirty.csv')

# Explore and Clean the Data

**Challenge**: Answer these questions about the dataset:
1. How many rows and columns does the dataset contain?
2. Are there any NaN values present?
3. Are there any duplicate rows?
4. What are the data types of the columns?

In [5]:
data.shape

(5391, 6)

In [7]:
data.isna().values.any()

False

In [10]:
data.duplicated().any()

False

In [12]:
data.dtypes

Rank                      int64
Release_Date             object
Movie_Title              object
USD_Production_Budget    object
USD_Worldwide_Gross      object
USD_Domestic_Gross       object
dtype: object

In [13]:
data.sample(10)

Rank Release_Date            Movie_Title USD_Production_Budget   
1917  5149   11/15/2002           Revolution#9              $350,000  \
4920  1121     6/3/2015                Chappie           $49,000,000   
1125   862    8/14/1998           The Avengers           $60,000,000   
4185  2484     7/9/2012  The Cold Light of Day           $20,000,000   
760   1734   12/23/1994    Legends of the Fall           $30,000,000   
1393  2106     3/3/2000    The Next Best Thing           $25,000,000   
5367   162    7/21/2017                Dunkirk          $150,000,000   
2195  1560    3/12/2004                 Closer           $35,000,000   
1838  3379    6/28/2002  Hey Arnold! The Movie           $10,000,000   
2167  3037    1/23/2004   The Butterfly Effect           $13,000,000   

     USD_Worldwide_Gross USD_Domestic_Gross  
1917              $9,118             $9,118  
4920        $105,002,056        $31,569,268  
1125         $48,585,416        $23,385,416  
4185         $25,361,206         $3,763,583  
760         $160,502,573        $66,502,573  
1393         $24,355,762        $14,983,572  
5367        $235,907,054       $104,641,608  
2195        $116,177,695        $33,987,757  
1838         $13,684,949        $13,684,949  
2167         $95,638,592        $57,924,679

### Data Type Conversions

**Challenge**: Convert the `USD_Production_Budget`, `USD_Worldwide_Gross`, and `USD_Domestic_Gross` columns to a numeric format by removing `$` signs and `,`. 
<br>
<br>
Note that *domestic* in this context refers to the United States.

In [28]:
chars_to_remove = [',', '$']
columns_to_clean = ['USD_Production_Budget', 
                    'USD_Worldwide_Gross',
                    'USD_Domestic_Gross']
 
for col in columns_to_clean:
    for char in chars_to_remove:
        # Replace each character with an empty string
        data[col] = data[col].astype(str).str.replace(char, "")
    # Convert column to a numeric data type
    data[col] = pd.to_numeric(data[col])

**Challenge**: Convert the `Release_Date` column to a Pandas Datetime type. 

In [32]:
data.Release_Date.dtype

dtype('O')

In [33]:
data.Release_Date = pd.to_datetime(data.Release_Date)
data.Release_Date.dtype

dtype('<M8[ns]')

### Descriptive Statistics

**Challenge**: 

1. What is the average production budget of the films in the data set?
2. What is the average worldwide gross revenue of films?
3. What were the minimums for worldwide and domestic revenue?
4. Are the bottom 25% of films actually profitable or do they lose money?
5. What are the highest production budget and highest worldwide gross revenue of any film?
6. How much revenue did the lowest and highest budget films make?

In [39]:
print(f"$ {round(data.USD_Production_Budget.mean(), 2)} is the average production budget of the films in the data set")
print(f"$ {round(data.USD_Worldwide_Gross.mean(),2)} is the average worldwide gross revenue of films")
print(f"$ {data['USD_Worldwide_Gross'].min()} is miniums worldwide revenue and $ {data['USD_Domestic_Gross'].min()} is minimum domestic revenue")


$ 31113737.58 is the average production budget of the films in the data set
$ 88855421.96 is the average worldwide gross revenue of films
$ 0 is miniums worldwide revenue and $ 0 is minimum domestic revenue


In [60]:
data.describe()

Rank                   Release_Date  USD_Production_Budget   
count 5,391.00                           5391               5,391.00  \
mean  2,696.00  2003-09-19 15:02:02.203672704          31,113,737.58   
min       1.00            1915-08-02 00:00:00               1,100.00   
25%   1,348.50            1999-12-02 12:00:00           5,000,000.00   
50%   2,696.00            2006-06-23 00:00:00          17,000,000.00   
75%   4,043.50            2011-11-23 00:00:00          40,000,000.00   
max   5,391.00            2020-12-31 00:00:00         425,000,000.00   
std   1,556.39                            NaN          40,523,796.88   

       USD_Worldwide_Gross  USD_Domestic_Gross  
count             5,391.00            5,391.00  
mean         88,855,421.96       41,235,519.44  
min                   0.00                0.00  
25%           3,865,206.00        1,330,901.50  
50%          27,450,453.00       17,192,205.00  
75%          96,454,455.00       52,343,687.00  
max       2,783,918,982.00      936,662,225.00  
std         168,457,757.00       66,029,346.27

In [66]:
data[data['USD_Production_Budget'] == data['USD_Production_Budget'].min()]

Rank Release_Date        Movie_Title  USD_Production_Budget   
2427  5391   2005-05-08  My Date With Drew                   1100  \

      USD_Worldwide_Gross  USD_Domestic_Gross  
2427               181041              181041

In [67]:
data[data['USD_Production_Budget'] == data['USD_Production_Budget'].max()]

Rank Release_Date Movie_Title  USD_Production_Budget   
3529     1   2009-12-18      Avatar              425000000  \

      USD_Worldwide_Gross  USD_Domestic_Gross  
3529           2783918982           760507625

# Investigating the Zero Revenue Films

**Challenge** How many films grossed $0 domestically (i.e., in the United States)? What were the highest budget films that grossed nothing?

In [81]:
zero_domestic = data[data["USD_Domestic_Gross"] == 0]
print(f"{len(zero_domestic)} films grossed domestically 0 $")
print(f"{zero_domestic[zero_domestic['USD_Production_Budget'] == zero_domestic['USD_Production_Budget'].max()]['Movie_Title'].iloc[0]} is highest budget film which grossed nothing")


512 films grossed domestically 0 $
Singularity is highest budget film which grossed nothing


**Challenge**: How many films grossed $0 worldwide? What are the highest budget films that had no revenue internationally?

In [82]:
zero_worldwide = data[data["USD_Worldwide_Gross"] == 0]
print(f"{len(zero_worldwide)} films grossed worldwide 0 $")
print(f"{zero_worldwide[zero_worldwide['USD_Production_Budget'] == zero_worldwide['USD_Production_Budget'].max()]['Movie_Title'].iloc[0]} is highest budget film which grossed nothing")


357 films grossed worldwide 0 $
Singularity is highest budget film which grossed nothing


In [80]:
zero_worldwide.sort_values("USD_Production_Budget", ascending=False)

Rank Release_Date              Movie_Title  USD_Production_Budget   
5388    96   2020-12-31              Singularity              175000000  \
5387   126   2018-12-18                  Aquaman              160000000   
5384   321   2018-09-03        A Wrinkle in Time              103000000   
5385   366   2018-10-08           Amusement Park              100000000   
5058   880   2015-11-12         The Ridiculous 6               60000000   
...    ...          ...                      ...                    ...   
4787  5371   2014-12-31     Stories of Our Lives                  15000   
3056  5374   2007-12-31              Tin Can Man                  12000   
4907  5381   2015-05-19         Family Motocross                  10000   
5006  5389   2015-09-29  Signed Sealed Delivered                   5000   
5007  5390   2015-09-29     A Plague So Pleasant                   1400   

      USD_Worldwide_Gross  USD_Domestic_Gross  
5388                    0                   0  
5387                    0                   0  
5384                    0                   0  
5385                    0                   0  
5058                    0                   0  
...                   ...                 ...  
4787                    0                   0  
3056                    0                   0  
4907                    0                   0  
5006                    0                   0  
5007                    0                   0  

[357 rows x 6 columns]

### Filtering on Multiple Conditions

In [84]:
international_releases = data.loc[(data.USD_Domestic_Gross == 0) & 
                                  (data.USD_Worldwide_Gross != 0)]
international_releases

Rank Release_Date              Movie_Title  USD_Production_Budget   
71    4310   1956-02-16                 Carousel                3380000  \
1579  5087   2001-02-11  Everything Put Together                 500000   
1744  3695   2001-12-31                 The Hole                7500000   
2155  4236   2003-12-31                  Nothing                4000000   
2203  2513   2004-03-31                The Touch               20000000   
...    ...          ...                      ...                    ...   
5340  1506   2017-04-14      Queen of the Desert               36000000   
5348  2225   2017-05-05        Chāi dàn zhuānjiā               23000000   
5360  4832   2017-07-03                Departure                1100000   
5372  1856   2017-08-25                Ballerina               30000000   
5374  4237   2017-08-25     Polina danser sa vie                4000000   

      USD_Worldwide_Gross  USD_Domestic_Gross  
71                   3220                   0  
1579                 7890                   0  
1744             10834406                   0  
2155                63180                   0  
2203              5918742                   0  
...                   ...                 ...  
5340              1480089                   0  
5348             58807172                   0  
5360                27561                   0  
5372             48048527                   0  
5374                36630                   0  

[155 rows x 6 columns]

**Challenge**: Use the [`.query()` function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.query.html) to accomplish the same thing. Create a subset for international releases that had some worldwide gross revenue, but made zero revenue in the United States. 

Hint: This time you'll have to use the `and` keyword.

In [93]:
international_releases = data.query("(@data.USD_Domestic_Gross == 0) and (@data.USD_Worldwide_Gross != 0)")
international_releases.tail(2)
# @ helps to access surrounding variables

Rank Release_Date           Movie_Title  USD_Production_Budget   
5372  1856   2017-08-25             Ballerina               30000000  \
5374  4237   2017-08-25  Polina danser sa vie                4000000   

      USD_Worldwide_Gross  USD_Domestic_Gross  
5372             48048527                   0  
5374                36630                   0

In [91]:
international_releases = data.query('USD_Domestic_Gross == 0 and USD_Worldwide_Gross != 0')
print(f'Number of international releases: {len(international_releases)}')
international_releases.tail()

Number of international releases: 155


Rank Release_Date           Movie_Title  USD_Production_Budget   
5340  1506   2017-04-14   Queen of the Desert               36000000  \
5348  2225   2017-05-05     Chāi dàn zhuānjiā               23000000   
5360  4832   2017-07-03             Departure                1100000   
5372  1856   2017-08-25             Ballerina               30000000   
5374  4237   2017-08-25  Polina danser sa vie                4000000   

      USD_Worldwide_Gross  USD_Domestic_Gross  
5340              1480089                   0  
5348             58807172                   0  
5360                27561                   0  
5372             48048527                   0  
5374                36630                   0

### Unreleased Films

**Challenge**:
* Identify which films were not released yet as of the time of data collection (May 1st, 2018).
* How many films are included in the dataset that have not yet had a chance to be screened in the box office? 
* Create another DataFrame called data_clean that does not include these films. 

In [ ]:
# Date of Data Collection
scrape_date = pd.Timestamp('2018-5-1')

### Films that Lost Money

**Challenge**: 
What is the percentage of films where the production costs exceeded the worldwide gross revenue? 

# Seaborn for Data Viz: Bubble Charts

### Plotting Movie Releases over Time

**Challenge**: Try to create the following Bubble Chart:

<img src=https://i.imgur.com/8fUn9T6.png>



# Converting Years to Decades Trick

**Challenge**: Create a column in `data_clean` that has the decade of the release. 

<img src=https://i.imgur.com/0VEfagw.png width=650> 

Here's how: 
1. Create a [`DatetimeIndex` object](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.html) from the Release_Date column. 
2. Grab all the years from the `DatetimeIndex` object using the `.year` property.
<img src=https://i.imgur.com/5m06Ach.png width=650>
3. Use floor division `//` to convert the year data to the decades of the films.
4. Add the decades as a `Decade` column to the `data_clean` DataFrame.

### Separate the "old" (before 1969) and "New" (1970s onwards) Films

**Challenge**: Create two new DataFrames: `old_films` and `new_films`
* `old_films` should include all the films before 1969 (up to and including 1969)
* `new_films` should include all the films from 1970 onwards
* How many films were released prior to 1970?
* What was the most expensive film made prior to 1970?

# Seaborn Regression Plots

**Challenge**: Use Seaborn's `.regplot()` to show the scatter plot and linear regression line against the `new_films`. 
<br>
<br>
Style the chart

* Put the chart on a `'darkgrid'`.
* Set limits on the axes so that they don't show negative values.
* Label the axes on the plot "Revenue in \$ billions" and "Budget in \$ millions".
* Provide HEX colour codes for the plot and the regression line. Make the dots dark blue (#2f4b7c) and the line orange (#ff7c43).

Interpret the chart

* Do our data points for the new films align better or worse with the linear regression than for our older films?
* Roughly how much would a film with a budget of $150 million make according to the regression line?

# Run Your Own Regression with scikit-learn

$$ REV \hat ENUE = \theta _0 + \theta _1 BUDGET$$

**Challenge**: Run a linear regression for the `old_films`. Calculate the intercept, slope and r-squared. How much of the variance in movie revenue does the linear model explain in this case?

# Use Your Model to Make a Prediction

We just estimated the slope and intercept! Remember that our Linear Model has the following form:

$$ REV \hat ENUE = \theta _0 + \theta _1 BUDGET$$

**Challenge**:  How much global revenue does our model estimate for a film with a budget of $350 million? 